### PR TITLE
Loading workerContext using the Knowledge manager

### LINKED ISSUE #2448: Loading context worker using the Knowledge manager
**Is your feature request related to a problem? Please describe.**
- Adding the ability to install multiple IGs into the worker context based on comment 1 [here](https://github.com/google/android-fhir/issues/2240#issuecomment-1774127729) 

> we could try to implement the IWorkerContext using knowledge manager, and essentially for knowledge manager to provide a worker context to the SDC library to use. This way, as an Android developer, your interation with npm packages should all go through knowledge manager.

**Would you like to work on the issue?**
@joiskash  and @nsabale7  will be looking into this

### PR DESCRIPTION
**Pull request by nsabale7**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2448

**Description**
Clear and concise code change description. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by owais-vd** [comment]
@nsabale7 just a question about the worker-context. I'm not sure why we created two different worker-context.
1. https://github.com/google/android-fhir/pull/2449/files#diff-0abdac65d1f5c5a9700253625ed1b9d36a465bc0b5fec7e7e9cf26c5abc02779R36
2. https://github.com/google/android-fhir/pull/2449/files#diff-63f547d1547f3d76757693a33c9f6ff6b911b03ab2aa4280843f9ddcddee3821R201

It shouldn't be one? and where the [loadWorkerContext](https://github.com/google/android-fhir/pull/2449/files#diff-63f547d1547f3d76757693a33c9f6ff6b911b03ab2aa4280843f9ddcddee3821R195) method is being called?

File: knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt [hunk 1]
```diff
@@ -189,6 +192,20 @@ internal constructor(
     return knowledgeDao.insertResource(igId, res)
   }
 
+  private suspend fun loadWorkerContext(
```
Comment (joiskash) [inline knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt:195]: This function is not being used anywhere. Is there something that is missing?

**Comment by nsabale7** [comment]
> 1. https://github.com/google/android-fhir/pull/2449/files#diff-0abdac65d1f5c5a9700253625ed1b9d36a465bc0b5fec7e7e9cf26c5abc02779R36
A. This defines a default simple worker-context object which does not load any packages.
> 2. https://github.com/google/android-fhir/pull/2449/files#diff-63f547d1547f3d76757693a33c9f6ff6b911b03ab2aa4280843f9ddcddee3821R201
A. This worker-context loads provided npmPackages and returns the object.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt
Comment (joiskash) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/mapping/ResourceMapperTest.kt:1]: Add a UT where multiple IGs are loaded.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt [hunk 2]
```diff
@@ -16,20 +16,24 @@
...
    * Optionally pass a custom version of [StructureMapUtilities.ITransformerServices] to support
    * specific use cases.
    */
   val transformSupportServices: StructureMapUtilities.ITransformerServices? = null,
+  /**
+   * Optionally pass a custom version of [IWorkerContext].
+   *
+   * @default [SimpleWorkerContext]
+   */
+  val simpleWorkerContext: IWorkerContext = SimpleWorkerContext(),
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt:36]: This context should be set through the knowledge manager if it is available.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 3]
```diff
@@ -200,9 +199,7 @@ object ResourceMapper {
   ): Bundle {
     val structureMapProvider = structureMapExtractionContext.structureMapProvider
     val simpleWorkerContext =
-      DataCapture.getConfiguration(structureMapExtractionContext.context)
-        .simpleWorkerContext
-        .apply { setExpansionProfile(Parameters()) }
+      structureMapExtractionContext.simpleWorkerContext.apply { setExpansionProfile(Parameters()) }
```
Comment (joiskash) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:202]: This simpleWorkerContext should be retrieved from the knowledge Manager

**Review by joiskash — CHANGES_REQUESTED** [review]
Please take a look at my comments. Thanks.

File: knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt (hunk 1 above)
Comment (nsabale7) [reply knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt:195]: Rhis is knowledge manager api which returns worker context and can be used the application.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt (hunk 2 above)
Comment (nsabale7) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt:36]: currently this has default value and can be replaced.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 3 above)
Comment (nsabale7) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:202]: Please refer [this ](https://github.com/google/android-fhir/issues/2448#issuecomment-1962293873). It describes the usage of this api

**Comment by jingtang10** [comment]
fyi left a comment in the issue: https://github.com/google/android-fhir/issues/2448#issuecomment-1979375776

i think we need to resolve this before we can merge this pr

**Comment by nsabale7** [comment]
@joiskash @owais-vd @dubdabasoduba @jingtang10 I have added test using outbreak package. Please review this PR.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt (hunk 3 above)
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt]: The simpleWorkerContext is being loaded from KM and provided via `StructureMapExtractionContext`. This is the agreed behaviour from Jing's comment [here](https://github.com/google/android-fhir/issues/2448#issuecomment-1983793669)

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt (hunk 2 above)
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/StructureMapExtractionContext.kt]: The workerContext is currnetly loaded using KM api `loadWorkerContext` and provided to the datacapture (specifically ResourceMapper) via `StructureMapExtractionContext`. If that's what you mean by "through" KM then we can resolve this comment.

**Review by MJ1998 — APPROVED** [review]
LGTM

File: knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt [hunk 4]
```diff
@@ -189,6 +192,30 @@ internal constructor(
     return knowledgeDao.insertResource(igId, res)
   }
 
+  /**
+   * Loads and initializes a worker context with the specified npm packages.
```
Comment (MJ1998) [inline knowledge/src/main/java/com/google/android/fhir/knowledge/KnowledgeManager.kt:196]: Add in this kdoc a use case of this api or link to the test in `ResourceMapperTest`.

**Pull request merged**